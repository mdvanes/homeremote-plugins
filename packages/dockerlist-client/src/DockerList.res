// {..} means we are handling a JS object with an unknown
// set of attributes
@module external styles: {..} = "./DockerList.module.css"

type request
type response

@new external makeXMLHttpRequest: unit => request = "XMLHttpRequest"
@send
external addEventListener: (request, string, unit => unit) => unit = "addEventListener"
@get external response: request => response = "response"
@send external open_: (request, string, string) => unit = "open"
@send external send: request => unit = "send"
@send external abort: request => unit = "abort"

@get external status: request => int = "status"

module DockerListMod = {
  type t = {
    address: string,
    total_balance: float,
  }

  @scope("JSON") @val
  external // external parseResponse: response => t = "parse"
  parseResponse: response => {"message": array<string>} = "parse"

  //  let query = (~address, ~onDone, ~onError, ()) => {
  let query = (~address) => {
    let request = makeXMLHttpRequest()

    // request->addEventListener("load", () => onDone(request->response->parseResponse))
    // request->addEventListener("error", () => onError(request->status))
    request->addEventListener("load", () => {
      let response = request->response->parseResponse
      Js.log(response["message"])
    })
    request->addEventListener("error", () => {
      Js.log("Error logging here")
    })

    request->open_("GET", "https://dog.ceo/api/breeds/image/random/" ++ address)
    request->send

    // () => request->abort
  }

  @react.component
  let make = (~count: int) => {
    let times = switch count {
    | 1 => "once"
    | 2 => "twice"
    | n => Belt.Int.toString(n) ++ " times"
    }
    let msg = "Click me " ++ times

    // Runs only once right after mounting the component
    React.useEffect0(() => {
      // Run effects
      query(~address="3")
      None // or Some(() => {})
    })

    <div
      style={ReactDOM.Style.make(
        // ~backgroundColor="white",
        ~padding="2px",
        ~borderRadius="2px",
        (),
      )}>
      {React.string("Docker List")}
      <button className={styles["root"]}> {msg->React.string} </button>
    </div>
  }
}

// Js.log("Hello, World!")
