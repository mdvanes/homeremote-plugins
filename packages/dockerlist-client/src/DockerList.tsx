import React, { useRef } from "react";
// import "./StreamPlayer.css";
// import Elm from "react-elm-components";
// import Audio from "./Elm/Audio.elm";

// const setupPorts = (containerRef) => (ports) => {
//     ports.setPlayPauseStatusPort.subscribe((newStatus) => {
//         // NOTE: findDomNode has been deprecated
//         const audioElem = containerRef.current.children[0].querySelector(
//             "audio"
//         );
//         // Wait to let the audio elem be updated with a new cachebusting timestamp in Audio.elm `Cmd.batch [ Task.perform UpdateTimestamp Time.now, Cmd.map MsgControls controlsCmds ]`
//         setTimeout(() => {
//             if (newStatus === "Play") {
//                 audioElem.play();
//             } else {
//                 audioElem.pause();
//             }
//         }, 50);
//     });
// };

const DockerList = ({ url }) => {
    const containerRef = useRef(null);
    return (
        <div ref={containerRef}>
            <div
                style={{
                    backgroundColor: "white",
                    padding: 2,
                    borderRadius: 2,
                }}
            >
                Docker List
            </div>

            {/* <Elm
                src={Audio.Elm.Elm.Audio}
                flags={{ url }}
                ports={setupPorts(containerRef)}
            /> */}
        </div>
    );
};

export default DockerList;
