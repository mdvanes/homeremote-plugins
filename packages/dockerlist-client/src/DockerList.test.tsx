import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { DockerListMod } from "./DockerList.bs";

const DockerList = DockerListMod.make;

jest.mock("./DockerList.module.css", () => "");

const MOCK_DOCKERLIST_RESPONSE = {
    status: "received",
    containers: [
        {
            Id:
                "2757091519a2e227b40e6fae4f92b7f4c8b7b189037bda0bc3152ce2be96af9d",
            Names: ["/determined_edison"],
            State: "running",
            Status: "Up 2 days",
        },
        {
            Id:
                "42aebf279f8c95488fab905d788f3caffee6afcaad240fc4aca68106c7173bfe",
            Names: ["/tinycors"],
            State: "exited",
            Status: "Exited (2) 13 hours ago",
        },
        {
            Id:
                "c7f51601f9870180859a2feb981613b30e3fcd282830b398185aa4facff21be1",
            Names: ["/hello_world"],
            State: "exited",
            Status: "Exited (0) 17 hours ago",
        },
    ],
};

window.fetch = jest.fn().mockResolvedValue({
    text: () => Promise.resolve(JSON.stringify(MOCK_DOCKERLIST_RESPONSE)),
});

describe("DockerList client", () => {
    it("renders buttons for each docker container", async () => {
        const { container } = render(<DockerList url={"some_url"} />);
        await screen.findByText("determined_edison");
        // const buttons = screen.getAllByRole("button");
        expect(screen.getAllByRole("button").length).toBe(3);
        // const buttonsParent = buttons[0].parentNode;
        expect(container).toMatchSnapshot("container");
    });

    it("can stop running container", async () => {
        render(<DockerList url={"some_url"} />);

        // Note: find instead of waitFor(await getByText("determined_edison"))
        await screen.findByText("determined_edison");

        const runningContainerButton = screen.getByRole("button", {
            name: "determined_edison Up 2 days",
        });

        userEvent.click(runningContainerButton);

        await screen.findByText(/Do you want to/);

        expect(
            screen.getByText(/Do you want to stop determined_edison/)
        ).toBeInTheDocument();

        const okButton = screen.getByRole("button", {
            name: "OK",
        });

        const expectedUrl =
            "some_url/api/dockerlist/stop/2757091519a2e227b40e6fae4f92b7f4c8b7b189037bda0bc3152ce2be96af9d";

        expect(window.fetch).not.toBeCalledWith(expectedUrl);
        userEvent.click(okButton);

        expect(
            await screen.findByText(/Do you want to/)
        ).not.toBeInTheDocument();

        expect(window.fetch).toBeCalledWith(expectedUrl);
    });

    it("can start stopped container", async () => {
        render(<DockerList url={"some_url"} />);

        // Note: find instead of waitFor(await getByText("determined_edison"))
        await screen.findByText("determined_edison");

        const stoppedContainerButton = screen.getByRole("button", {
            name: "hello_world Exited (0) 17 hours ago",
        });

        userEvent.click(stoppedContainerButton);

        await screen.findByText(/Do you want to/);

        expect(
            screen.getByText(/Do you want to start hello_world/)
        ).toBeInTheDocument();

        const okButton = screen.getByRole("button", {
            name: "OK",
        });

        const expectedUrl =
            "some_url/api/dockerlist/start/c7f51601f9870180859a2feb981613b30e3fcd282830b398185aa4facff21be1";

        expect(window.fetch).not.toBeCalledWith(expectedUrl);
        userEvent.click(okButton);

        expect(
            await screen.findByText(/Do you want to/)
        ).not.toBeInTheDocument();

        expect(window.fetch).toBeCalledWith(expectedUrl);
    });
});
