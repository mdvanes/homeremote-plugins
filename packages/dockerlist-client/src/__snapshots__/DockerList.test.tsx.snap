// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DockerList client renders buttons for each docker container: container 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <div
          class="MuiBox-root MuiBox-root-7"
        />
        <div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemIcon-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-3 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-4 Mui-checked MuiIconButton-colorSecondary MuiIconButton-edgeStart"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <input
                    aria-labelledby="2757091519a2e227b40e6fae4f92b7f4c8b7b189037bda0bc3152ce2be96af9d"
                    checked=""
                    class="PrivateSwitchBase-input-6"
                    data-indeterminate="false"
                    type="checkbox"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </span>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
              id="2757091519a2e227b40e6fae4f92b7f4c8b7b189037bda0bc3152ce2be96af9d"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
              >
                determined_edison
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                Up 2 days
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
        </div>
      </ul>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <div
          class="MuiBox-root MuiBox-root-2"
        />
        <div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemIcon-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-3 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary MuiIconButton-edgeStart"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <input
                    aria-labelledby="c7f51601f9870180859a2feb981613b30e3fcd282830b398185aa4facff21be1"
                    class="PrivateSwitchBase-input-6"
                    data-indeterminate="false"
                    type="checkbox"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </span>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
              id="c7f51601f9870180859a2feb981613b30e3fcd282830b398185aa4facff21be1"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
              >
                hello_world
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                Exited (0) 17 hours ago
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
        </div>
        <div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemIcon-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-3 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary MuiIconButton-edgeStart"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <input
                    aria-labelledby="42aebf279f8c95488fab905d788f3caffee6afcaad240fc4aca68106c7173bfe"
                    class="PrivateSwitchBase-input-6"
                    data-indeterminate="false"
                    type="checkbox"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </span>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
              id="42aebf279f8c95488fab905d788f3caffee6afcaad240fc4aca68106c7173bfe"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
              >
                tinycors
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                Exited (2) 13 hours ago
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
        </div>
      </ul>
    </div>
  </div>
</div>
`;
